### PR TITLE
no network

### LINKED ISSUE #415: give useful error / warning message when device not connected to the internet
right now, we just quietly don't load anything :(

Summary of quick online/offline/online toggle testing from @devinreams:

1. blank white webviews do not contain useful "offline" errors or messages

2. depending on connectivity loss timing: a user could get really stuck after confirmation in "no entries" and not see their logins load in even after connectivity restored

### Visual Design

**InVision**
WebViews: https://mozilla.invisionapp.com/share/BQEUSO0S2#/311856390_No_Internet_-_WebView
FxA: https://mozilla.invisionapp.com/share/BQEUSO0S2#/311856391_No_Internet_-_FxA
Unpopulated Entry List: https://mozilla.invisionapp.com/share/BQEUSO0S2#/311856389_No_Internet_-_Entry_List_-_Not_Populated
Populated Entry List: https://mozilla.invisionapp.com/share/BQEUSO0S2#/311856388_No_Internet_-_Entry_List_-_Populated

**Zeplin**
WebViews: https://zpl.io/agvo46M
FxA: https://zpl.io/VOkMwA1
Unpopulated Entry List: https://zpl.io/bLO8l1G
Populated Entry List: https://zpl.io/Vk5NRWM

### Engineering Checklist
- [x] ItemList
- [x] WebView
- [x] FxA
- [x] unlock works on no network

### LINKED ISSUE #1004: Unable to unlock app without internet connection
Build: 1.6.0(3339)
Device: iPhone Xr
iOS: 12.0

Precondition
- Biometrics to unlock enabled

Steps to reproduce:
1.  Lock and close the app
2. Turn off internet connection
3. Launch the app
4. Tap the Unlock button

Actual results:
- Unlocking the app is not possible without internet connection. The "Unlock" button does not prompt biometrics to unlock the app.

### PR DESCRIPTION
**Pull request by sashei**
Fixes #415 
Fixes #1004 

## Testing and Review Notes

I tried to test cases not only related to being disconnected, but also around the boundaries of being disconnected and then reconnecting.

## To Do

- [x] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)

### CODE REVIEW

File: Shared/Action/NetworkAction.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+enum NetworkAction: Action {
+    case retry
+}
+
+extension NetworkAction: Equatable {
+    static func ==(lhs: NetworkAction, rhs: NetworkAction) -> Bool {
+        switch (lhs, rhs) {
+        case (.retry, .retry): return true
+        }
```
Comment (jhugman) [inline Shared/Action/NetworkAction.swift]: Can you omit the `return false` now? Wow.

File: Shared/Presenter/BaseItemListPresenter.swift [hunk 2]
```diff
@@ -21,6 +21,7 @@ struct LoginListTextSort {
     let sortingOption: Setting.ItemListSort
     let syncState: SyncState
     let storeState: LoginStoreState
+    let networkConnected: Bool
```
Comment (jhugman) [inline Shared/Presenter/BaseItemListPresenter.swift]: At some point this `struct` should change name.

`LoginListTextSort` doesn't really capture why it should have `networkConnected` in it.

File: Shared/Presenter/BaseItemListPresenter.swift [hunk 3]
```diff
@@ -64,25 +67,26 @@ class BaseItemListPresenter {
...
         }.asObserver()
     }()
 
-    lazy fileprivate var emptyPlaceholderItems = [
-        ItemSectionModel(model: 0, items: [LoginListCellConfiguration.EmptyListPlaceholder(learnMoreObserver: self.learnMoreObserver)]
-        )
-    ]
+    lazy private(set) var retryNetworkObserver: AnyObserver<Void> = {
+        return Binder(self) { target, _ in
+            target.dispatcher.dispatch(action: NetworkAction.retry)
```
Comment (jhugman) [inline Shared/Presenter/BaseItemListPresenter.swift]: Nit: whitespace

File: Shared/Presenter/BaseItemListPresenter.swift [hunk 4]
```diff
@@ -64,25 +67,26 @@ class BaseItemListPresenter {
...
+    lazy fileprivate var emptyPlaceholderItems = [LoginListCellConfiguration.EmptyListPlaceholder(learnMoreObserver: self.learnMoreObserver)]
+
+    lazy fileprivate var noResultsPlaceholderItems = [LoginListCellConfiguration.NoResults(learnMoreObserver: self.learnMoreNewEntriesObserver)]
+
+    lazy internal var syncPlaceholderItems = [LoginListCellConfiguration.SyncListPlaceholder]
 
-    lazy internal var syncPlaceholderItems = [
-        ItemSectionModel(model: 0, items: [LoginListCellConfiguration.SyncListPlaceholder])
-    ]
+    lazy internal var noNetworkItems = [LoginListCellConfiguration.NoNetwork(retryObserver: self.retryNetworkObserver)]
```
Comment (jhugman) [inline Shared/Presenter/BaseItemListPresenter.swift]: Does this mean that when the network isn't available, then neither are your passwords? That doesn't seem very useful.

File: Shared/Presenter/BaseItemListPresenter.swift [hunk 5]
```diff
@@ -157,27 +165,31 @@ extension BaseItemListPresenter {
...
+                    detailItemId: latest.7 ? latest.6 : "" // only pass along the detailItemId if we are showing the sidebar
                 )
             }
             .distinctUntilChanged()
             .map { (latest: LoginListTextSort) -> [ItemSectionModel] in
+                let networkItems = latest.networkConnected ? [] : self.noNetworkItems
+                
                 if latest.syncState == .Syncing && latest.logins.isEmpty {
-                    return self.syncPlaceholderItems
+                    return [ItemSectionModel(model: 0, items: networkItems + self.syncPlaceholderItems)]
```
Comment (jhugman) [inline Shared/Presenter/BaseItemListPresenter.swift]: Oh, I see.

File: Shared/Store/NetworkStore.swift [hunk 6]
```diff
@@ -41,5 +47,12 @@ open class NetworkStore {
         if !isRunningTest {
             ((try? self.reachability?.startNotifier()) as ()??)
         }
+
+        dispatcher.register
+                .filterByType(class: NetworkAction.self)
+                .filter { $0 == NetworkAction.retry }
+                .map { _ in (self.reachability?.connection ?? Reachability.Connection.none) != .none }
+                .bind(to: self._connectedToNetwork)
+                .disposed(by: self.disposeBag)
```
Comment (jhugman) [inline Shared/Store/NetworkStore.swift]: I think this is good, but doesn't overcome the limitations of the reachability API.

File: Shared/View/BaseItemListView.swift [hunk 7]
```diff
@@ -247,6 +247,16 @@ extension BaseItemListView {
                         borderView.backgroundColor = Constant.color.helpTextBorderColor
                         cell.addSubview(borderView)
 
+                        retCell = cell
+                    case .NoNetwork(let retryObserver):
+                        guard let cell = tableView.dequeueReusableCell(withIdentifier: "nonetwork") as? NoNetworkCell else {
+                            fatalError("couldn't find the nonetwork cell")
```
Comment (jhugman) [inline Shared/View/BaseItemListView.swift]: Ug.

File: Shared/Presenter/BaseItemListPresenter.swift (hunk 4 above)
Comment (sashei) [reply Shared/Presenter/BaseItemListPresenter.swift]: no, you'll still see the items show up in _addition_ to the no network disclaimer item.

File: Shared/View/BaseItemListView.swift (hunk 7 above)
Comment (sashei) [reply Shared/View/BaseItemListView.swift]: if only there was a better way...

File: Shared/Store/NetworkStore.swift (hunk 6 above)
Comment (sashei) [reply Shared/Store/NetworkStore.swift]: in practice, the `retry` action is rarely required; the callbacks preceding this subscription actually get fired very quickly!

**Pull request merged**